GameServerLib/Content/NavGrid.cs

```diff
@@ -54,22 +43,106 @@ public List<Vector2> GetPath(Vector2 from, Vector2 to)
+                while (true)
                 {
-                    foreach (var position in path)
+                    if (!priorityQueue.TryFirst(out path))
+                    {
+                        // no solution
+                        path = null;
+                        break;
+                    }
+                    var curCost = priorityQueue.GetPriority(priorityQueue.First);
+                    priorityQueue.TryDequeue(out path);
+                    var cell = path.Peek();
+                    curCost -= (CellDistance(cell, goal) + cell.Heuristic); // decrease the heuristic to get the cost
+                    // found the min solution return it (path)
+                    if (cell.Id == goal.Id) break;
+                    NavGridCell tempCell = null;
+                    foreach (NavGridCell ncell in GetCellNeighbors(cell))
                     {
-                        var navGridCell = GetCell(position.X, position.Y);
-                        var cellPosition = TranslateFromNavGrid(new Vector<float>() { X = navGridCell.X, Y = navGridCell.Y });
-                        returnList.Add(new Vector2(cellPosition.X, cellPosition.Y));
+                        // if the neighbor in the closed list - skip
+                        if (closedList.TryGetValue(ncell.Id, out tempCell)) continue;
+                        // not walkable - skip
+                        if (ncell.HasFlag(this, NavigationGridCellFlags.NOT_PASSABLE) || ncell.HasFlag(this, NavigationGridCellFlags.SEE_THROUGH))
+                        {
+                            closedList.Add(ncell.Id, ncell);
+                            continue;
+                        }
+                        // calculate the new path and cost +heuristic and add to the priority queue
+                        var npath = new Stack<NavGridCell>(new Stack<NavGridCell>(path));
+                        npath.Push(ncell);
+                        // add 1 for every cell used
+                        priorityQueue.Enqueue(npath, curCost + 1 + ncell.Heuristic + ncell.ArrivalCost + ncell.AdditionalCost + CellDistance(ncell, goal));
+                        closedList.Add(ncell.Id, ncell);
                     }
                 }
+                // shouldn't happen usually
+                if (path == null) return null;
+                var arr = path.ToArray();
+                Array.Reverse(arr);
+                var pathList = SmoothPath(new List<NavGridCell>(arr));
+                pathList.RemoveAt(0); // removes the first point
+                foreach (var navGridCell in pathList.ToArray())
+                {
+                    var cellPosition = TranslateFromNavGrid(new Vector<float>() { X = navGridCell.X, Y = navGridCell.Y });
+                    returnList.Add(new Vector2(cellPosition.X, cellPosition.Y));
+                }
+                return returnList;
+            }
+            //Logging.LoggerProvider.GetLogger().Debug("Found path: "+returnList.ToString());
```

Comment:
Looks like you forgot a leftover debugging line

Reply:
Removed.